lib/vector-core/buffers/benches/on_disk.rs

```diff
@@ -0,0 +1,198 @@
+use buffers::{self, Variant, WhenFull};
+use criterion::{
+    criterion_group, criterion_main, measurement::WallTime, BatchSize, BenchmarkGroup, BenchmarkId,
+    Criterion, SamplingMode, Throughput,
+};
+use std::mem;
+use std::sync::atomic::AtomicUsize;
+use std::time::Duration;
+use std::{path::PathBuf, sync::atomic::Ordering};
+
+mod common;
+
+const ROTATION_THRESHOLD: usize = 10_000;
+
+/// A struct to manage the data_dir of an on-disk benchmark
+///
+/// The way our benchmarks function we have the choice of sharing a data_dir
+/// between benchmarks or give each benchmark its own data_dir. The first option
+/// will cause cross-pollution of benchmarks so that's a no-go. The second
+/// option will fill up the benchmarker's disk. Filling up disk is a no-go but
+/// is something we can manage.
+///
+/// This struct keeps track of the current iteration of the tests and manages
+/// new sub-data_dir paths. These paths, wrapped in a [`PathGuard`]
+/// self-destruct when the benchmark drops them while this struct self-destructs
+/// when it is dropped. This keeps disk consumption to a minimum.
+struct DataDir {
+    counter: AtomicUsize,
+    base: PathBuf,
+}
+
+struct PathGuard {
+    inner: PathBuf,
+}
+
+impl DataDir {
+    fn new(name: &str) -> Self {
+        let mut base_dir = PathBuf::new();
+        base_dir.push(std::env::temp_dir());
+        base_dir.push(name);
```

Comment:
Note that this can be done in one step, something like:
```rust
let base_dir: PathBuf = [std::env::temp_dir().as_str(), name].iter().collect();
```